File: code_samples/go/functional_map_reduce.go
package main

import (
    "fmt"
)

// Functional: Map Reduce
// AI/ML Training Sample

type MapReduce struct {
    Data string
}

func NewMapReduce() *MapReduce {
    return &MapReduce{
        Data: "",
    }
}

func (s *MapReduce) Process(input string) {
    s.Data = input
}

func (s *MapReduce) Validate() bool {
    return len(s.Data) > 0
}

func (s *MapReduce) GetData() string {
    return s.Data
}

func main() {
    instance := NewMapReduce()
    instance.Process("example")
    fmt.Printf("Data: %s\n", instance.GetData())
    fmt.Printf("Valid: %v\n", instance.Validate())
}

File: code_samples/haskell/functional_map_reduce.hs
-- Functional: Map Reduce
-- AI/ML Training Sample

module MapReduce where

data MapReduce = MapReduce {
    getData :: String
} deriving (Show, Eq)

process :: MapReduce -> String -> MapReduce
process obj input = obj { getData = input }

validate :: MapReduce -> Bool
validate obj = not (null (getData obj))

main :: IO ()
main = do
    let instance = MapReduce { getData = "" }
    let updated = process instance "example"
    putStrLn $ "Data: " ++ getData updated
    putStrLn $ "Valid: " ++ show (validate updated)
